components/script/dom/idbfactory.rs

```diff
@@ -0,0 +1,227 @@
+use std::sync::mpsc;
+use std::thread::Builder;
+
+use servo_url::origin::ImmutableOrigin;
+
+use crate::script_runtime::JSContext as SafeJSContext;
+use js::rust::HandleValue;
+
+#[dom_struct]
+pub struct IDBFactory {
+    reflector_: Reflector,
+}
+
+impl IDBFactory {
+    pub fn new_inherited() -> IDBFactory {
+        IDBFactory {
+            reflector_: Reflector::new(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> DomRoot<IDBFactory> {
+        reflect_dom_object(
+            Box::new(IDBFactory::new_inherited()),
+            global,
+            IDBFactoryBinding::Wrap,
+        )
+    }
+
+    fn open_internal(
+        &self,
+        name: DOMString,
+        version: Option<u64>,
+        request: &IDBOpenDBRequest,
+        upgrade_notificator: mpsc::Sender<Result<(), ()>>,
+        result_channel: mpsc::Sender<Result<Trusted<IDBDatabase>, Error>>,
+    ) {
+        let global = self.global();
+        let window = global.as_window();
+
+        let result = IDBDatabase::new(window, name, version);
+
+        let db_version = result.version();
+        let new_version = match version {
+            Some(v) => v,
+            None => {
+                // db_version == 0 equals database == null
+                if db_version == 0 {
+                    1
+                } else {
+                    db_version
+                }
+            },
+        };
+
+        if new_version == db_version {
+            upgrade_notificator.send(Ok(())).unwrap();
+            result_channel.send(Ok(Trusted::new(&result))).unwrap();
+        } else if new_version > db_version {
+            // Step 5.10.4
+            // FIXME:(rasviitanen) Do this for all open handles
+            result.dispatch_versionchange(db_version, version);
+
+            // Step 5.10.6
+            // Dispatch an event to trigger version change
+            let transaction = IDBTransaction::new(
+                &self.global().as_window(),
+                &result,
+                IDBTransactionMode::Versionchange,
+                result.object_stores(),
+            );
+
+            transaction.set_active_flag(false);
+            transaction.start_and_upgrade_version(new_version); // Starts transaction
+
+            result.set_transaction(&transaction);
+
+            request.dispatch_upgrade_transaction(
+                &result,
+                &transaction,
+                db_version,
+                Some(new_version),
+                upgrade_notificator,
+            );
+
+            result_channel.send(Ok(Trusted::new(&result))).unwrap();
+        } else {
+            upgrade_notificator.send(Ok(())).unwrap();
+            result_channel.send(Err(Error::Version)).unwrap();
+        }
+    }
+}
+
+impl IDBFactoryMethods for IDBFactory {
+    // https://www.w3.org/TR/IndexedDB-2/#dom-idbfactory-open
+    #[allow(unsafe_code)]
+    fn Open(&self, name: DOMString, version: Option<u64>) -> Fallible<DomRoot<IDBOpenDBRequest>> {
```

Comment:
Ok so to be a bit clearer on what I meant. Here is how you could consider doing this:

1. let `request` be a new `IDBOpenDBRequest`.
2. create an ipc channel.
3. Setup a route with the receiver of that channel.
4. Send the sender inside a message to the indexDB component.
5. Create some sort of struct like:
```rust
struct OpenListener {
    open_request: Trusted<`IDBOpenDBRequest`>
    global: Trusted<Global>,
}

impl OpenListener {
    fn handle_upgrade_transaction_started() {
        // queue a task to:
        // 1. Set open_request’s result to connection.
        // 2. Set open_request’s transaction to transaction.
       // 3. fire a version change event 
       // and more from Step 9 of http://w3c.github.io/IndexedDB/#run-an-upgrade-transaction
    }
    fn handle_database_opened() {
        // queue a task to Set request’s done flag to true, and so on...
       // and more from http://w3c.github.io/IndexedDB/#ref-for-open-a-database%E2%91%A0
    }
}
```

And then from the "in-parallel" steps running in the `net` component, you can send the various messages over IPC back to script using the sender sent in the initial message from script, and the `OpenListener` on the `ROUTER` thread would handle messages by calling it's appropriate methods, which would always queue a task as per the spec to manipulate that various DOM objects and fire events.